### PR TITLE
Remove the use of euclid from the Window API

### LINKED ISSUE #1542: Remove re-export of `euclid` from public API
In the API review, we decided to replace the re-export of euclid types and their use in the public `Window` API in favor of our own API.

### PR DESCRIPTION
**Pull request by tronical**
Closes #1542

### CODE REVIEW

File: internal/core/api.rs [hunk 1]
```diff
@@ -12,10 +12,152 @@ use alloc::boxed::Box;
 use crate::component::ComponentVTable;
 use crate::window::{WindowAdapter, WindowInner};
 
-pub use crate::lengths::LogicalPx;
-pub use crate::lengths::PhysicalPx;
+/// A position represented in the coordinate space of logical pixels. That is the space before applying
+/// a display device specific scale factor.
+#[derive(Debug, Default, Copy, Clone, PartialEq)]
+#[repr(C)]
```
Comment (tronical) [inline internal/core/api.rs:18]: Note to self: revisit this

File: examples/mcu-board-support/Cargo.toml [hunk 2]
```diff
@@ -17,7 +17,7 @@ links = "mcu_board_support" # just so we can pass metadata to the slint build cr
 path = "lib.rs"
 
 [features]
-pico-st7789 = ["slint/unsafe-single-threaded", "rp-pico", "embedded-hal", "cortex-m-rt", "alloc-cortex-m", "fugit", "cortex-m", "cortex-m", "display-interface", "st7789", "defmt", "defmt-rtt",  "slint/defmt", "shared-bus", "slint/libm", "embedded-dma", "embedded-graphics"]
+pico-st7789 = ["slint/unsafe-single-threaded", "rp-pico", "embedded-hal", "cortex-m-rt", "alloc-cortex-m", "fugit", "cortex-m", "cortex-m", "display-interface", "st7789", "defmt", "defmt-rtt",  "slint/defmt", "shared-bus", "slint/libm", "embedded-dma", "embedded-graphics", "euclid/libm"]
```
Comment (ogoffart) [inline examples/mcu-board-support/Cargo.toml:20]: slint/libm hould imply euclid/libm 

This crate shouldn't depend on Euclid. Or what is the reason it does?

File: internal/core/api.rs [hunk 3]
```diff
@@ -264,25 +406,22 @@ pub use crate::input::PointerEventButton;
...
-    PointerMoved { position: euclid::Point2D<f32, LogicalPx> },
+    PointerMoved { position: LogicalPosition },
     /// Wheel was rotated.
     /// `pos` is the position of the mouse when the event happens.
     /// `delta` is the amount of pixel to scroll.
-    PointerScrolled {
-        position: euclid::Point2D<f32, LogicalPx>,
-        delta: euclid::Vector2D<f32, LogicalPx>,
-    },
+    PointerScrolled { position: LogicalPosition, delta: LogicalPosition },
```
Comment (ogoffart) [inline internal/core/api.rs:417]: I am a bit unsure about the use of a position type for the delta.
Maybe we could split into two f32 (horizontal and vertical delta) ?

File: internal/core/api.rs [hunk 4]
```diff
@@ -198,37 +340,37 @@ impl Window {
 
     /// This function returns an euclid scale that allows conveniently converting between logical and
     /// physical pixels based on the window's scale factor.
-    pub fn scale_factor(&self) -> euclid::Scale<f32, LogicalPx, PhysicalPx> {
-        self.0.scale()
+    pub fn scale_factor(&self) -> f32 {
```
Comment (ogoffart) [inline internal/core/api.rs:343]: I'm wondering if this should not return a newtype that can be used to multiply or divide the other types

File: examples/mcu-board-support/Cargo.toml (hunk 2 above)
Comment (tronical) [reply examples/mcu-board-support/Cargo.toml:20]: The XPT2046 uses it, for example `let mut point = Point2D::new(0u32, 0u32);` and `point += euclid::vec2(i16::MAX as u32 - x, y)`. I thought about changing that code as well, but figured it's out of scope for what is primarily an API change. I can do it though if you'd prefer.

File: internal/core/api.rs (hunk 3 above)
Comment (tronical) [reply internal/core/api.rs:417]: That's fine with me as well. I'm not particularly attached to using a struct here. I like separate fields.

File: internal/core/api.rs (hunk 4 above)
Comment (tronical) [reply internal/core/api.rs:343]: I thought about this as well but I must admit that I've come to really like this API, for two reasons:

1. It turns out that there are still many call sites that want just the scalar value (as you can see in this diff).
2. The conversion between logical and physical coordinates becomes more expressive:

```rust
let size_to_save_and_restore = window.size() / window.scale_factor();

// vs.

let size_to_save_and_restore = window.size().to_logical(window.scale_factor);
```

The `to_logical` is by far more readable that a single character `/`, in a context where it helps to minimise the effort to clarify whether the size variable is in logical or physical coordinate space.

File: internal/core/api.rs (hunk 3 above)
Comment (tronical) [reply internal/core/api.rs]: Done

File: examples/mcu-board-support/Cargo.toml (hunk 2 above)
Comment (ogoffart) [reply examples/mcu-board-support/Cargo.toml:20]: ah ok, then let's leave it for now

File: internal/core/api.rs [hunk 5]
```diff
@@ -264,25 +405,23 @@ pub use crate::input::PointerEventButton;
...
     /// Wheel was rotated.
     /// `pos` is the position of the mouse when the event happens.
-    /// `delta` is the amount of pixel to scroll.
-    PointerScrolled {
-        position: euclid::Point2D<f32, LogicalPx>,
-        delta: euclid::Vector2D<f32, LogicalPx>,
-    },
+    /// `delta_x` is the amount of pixels to scroll in horizontal direction,
+    /// `delta_y` is the amount of pixels to scroll in vertical direction.
+    PointerScrolled { position: LogicalPosition, delta_x: f32, delta_y: f32 },
```
Comment (ogoffart) [inline internal/core/api.rs:417]: how about `horizontal_delta` and `vertical_delta` ?  or do you think it is too verbose?

Comment (tronical) [reply internal/core/api.rs:417]: I was wondering the same. What makes me think that x/y is better is that it combines better with a z in the future, as opposed to a … depth?

Comment (tronical) [reply internal/core/api.rs:417]: The other source I was looking at was the DOM wheel event, which uses `deltaX`, `deltaY` and `deltaZ`: https://developer.mozilla.org/en-US/docs/Web/API/Element/wheel_event

**Review by ogoffart — APPROVED** [review]

**Comment by tronical** [comment]
Force-push was squash of the smaller fixes on top.

**Pull request merged**